Modules/_xxsubinterpretersmodule.c

```diff
@@ -2323,6 +2353,74 @@ PyDoc_STRVAR(channel_list_all_doc,
 \n\
 Return the list of all IDs for active channels.");
 
+
+static PyObject *
+channel_list_interpreters(PyObject *self, PyObject *args, PyObject *kwds)
+{
+    static char *kwlist[] = {"cid", "send", "recv", NULL};
+    int64_t cid;
+    _PyChannelState *chan;
+    int send = 0;           /* Send end? */
+    int recv = 0;           /* Receive end? */
+    int64_t *ids = NULL;    /* Array of interpreter IDs */
+    int64_t count = 0;      /* Number of interpreters to return */
+    PyObject *ret = NULL;   /* Python list of interpreter IDs */
+
+    if (!PyArg_ParseTupleAndKeywords(
+            args, kwds, "O&|$pp:channel_list_interpreters",
+            kwlist, channel_id_converter, &cid, &send, &recv)) {
+        return NULL;
+    }
+
+    if ((send && recv) || (!send && !recv)) {
+        PyErr_SetString(PyExc_ValueError,
+                        "Specify exactly one of send or recv");
+        goto except;
+    }
+
+    chan = _channels_lookup(&_globals.channels, cid, NULL);
+    if (chan == NULL) {
+        goto except;
+    }
+
+    ids = _channelends_list_interpreters(chan->ends, &count, send);
+    if (ids == NULL) {
+        goto except;
+    }
+
+    ret = PyList_New((Py_ssize_t)count);
+    if (ret == NULL) {
+        goto except;
```

Comment:
Let's keep this simple:

```suggestion
        PyMem_Free(ids);
        return NULL;
```

Reply:
While this might be simpler, I'm not sure I consider this better style. In the case above where `ids == NULL` there's no cleanup to do, so it seems ok to just return there.

However, as soon as any cleanup is involved it feels safer to have a single place where that's done. If there were more branches than this one catching errors (possibly added later on), each of them would have to at least do `PyMem_Free(ids); return NULL;`, rather than just `goto except;`.

For consistency I'd then advocate for keeping my existing structure of always using `except` and not having to check whether cleanup is needed in each of the error branches, with all of the cleanup handled in the except block. I'm mainly reluctant to duplicate any of the cleanup logic though (especially when the cleanup is unrelated to the error being caught).

Thoughts?

Reply:
I've realised if you go with consistent cleanup routines you also need any variables that are cleaned up to be declared at the top of the function rather than inline :) In this case it's `ret` that needs to be initialised to `NULL` ahead of the first `goto except`. I'm starting to see more of the reasoning behind the coding standards we use at work!

If you still want me to follow your suggestions I'm not going to push back - just thought I'd point out an argument for the structure I was using!

Reply:
Thanks for the explanation.  FWIW, I'm not C expert.  Nearly 100% of my C experience comes from working on Python. :)  So my style preferences are mostly derived from the CPython code base.  My suggestions stem from that.  Also, PEP 7 has the official style guidelines.

Anyway, I find what I suggested to be more readable, but I get your argument about guarding against future changes.  I tend to favor readability, but I'll leave it up to you. :)

Reply:
I've had another look at this, and it feels complicated enough of an example where I find it cleaner to keep the generic cleanup common at the end, rather than having to do the right set of cleanup in each of the three branches that use `goto except`.